### ISSUE TITLE
Add plugin system

### ISSUE THREAD

**Issue by goliatone** [goliatone/envset#1]

**Comment by moimikey** [comment]
@goliatone do you have some examples of what would constitute a viable plugin for envset?

**Comment by goliatone** [comment]
@moimikey This was meant as an early reminder of a feature I thought it would make sense to address an two issues I had on early development:
- syncing the `envset` file from a central repository- i.e. s3
- support `json` and `yaml` `envset` files

I did not start working on the plugin implementation for `envset`, but I have a working system in a different project. It's a little bit convoluted and was thinking on abstracting everything into it's own module.

Do you have any ideas or reference implementations of plugin architectures?

**Comment by moimikey** [comment]
@goliatone oh my, that sounds amazing, especially the distributed config. mucho 12-factor. i also +1 the idea of multi-language configuration.

Regarding multi-language configuration, `eslint` would be a possible example. Here's a particular part of their code base that determines if a particular config is present; it either has an extension or just `.eslintrc`; seemingly the project wants to avoid the non-file extension ways and considers it "legacy," which defaults to yaml. -- verbosity is always great, but at least there's the option to go extensionless: https://github.com/eslint/eslint/blob/720686bcf929b8a5331682339c6cc2c2de02afbd/lib/config/config-file.js#L51

Regarding pluggable interfaces, I don't blame you for considering your plugin implementation as convoluted, as I think most implementations are somewhat convoluted and specific to their use case. `eslint` has a very pluggable interface, but it's so specific to `eslint`. I think `nconf` would be an excellent example of a very agnostic pluggable interface: https://github.com/indexzero/nconf/blob/master/lib/nconf/provider.js. An example provider is the `nconf-redis` package to enable `nconf` to store in `redis`: https://github.com/indexzero/nconf-redis/blob/master/lib/nconf-redis.js

Winston is another example with their naming convention being `transports` vs `provider`: https://github.com/winstonjs/winston/tree/d567bfed171cf2074b6b07423a314a61db76e039/lib/winston/transports

I'd love to contribute, but I'll also want to do some research as well :)

**Comment by goliatone** [comment]
@moimikey My main concern regarding a plugin architecture is being able to manage- i.e. install- them in a transparent way. The implementation I mentioned before uses Commander as a vehicle. Basically you would `npm` install and the plugin would be available as a subcommand.

One of the things I enjoy about `envset` is it's simplicity- in general- but also in it's CLI API:

```
envset <environenment> -- <command>
```

I would like it to remain this way. Also, I started [porting envset to golang][1]. Ideally I would keep parity between both implementations. Keeping a terse API simplifies this. 

[1] https://github.com/goliatone/go-envset

**Comment by moimikey** [comment]
@goliatone got it. i can admire the minimalism. i've turned to using `yargs` which has a really nice api for sub-commands can become super pluggable. thoughts on yargs?

**edit** http://yargs.js.org/docs/#methods-commanddirdirectory-opts can imagine that we have similar code to this example which regex matches `node_modules` for `envset-plugin-*` and loads them in dynamically.

**Comment by goliatone** [comment]
@moimikey yes, I've used `yargs` before and I like it,  at work we were using `commander` and I just used it also for my personal projects due to inertia. `yargs` looks like it could be a good way to go about plugins tho

**Comment by moimikey** [comment]
i had used `commander` as well and then finally fell in love with the `commandDir` option in `yargs` to use subdirectories as subcommands. such a seamly way to decouple.

So definitely +1 for `yargs`. i'd love to contribute in any way i can, as I use `envset` on work/personal projects and would love the idea of having the option to add a distributed store with-or-without atomic updates.

💃 💃 💃

**Comment by moimikey** [comment]
we can also have some form of dynamic requires. if for say there are these two made up plugins:

`envset-plugin-s3`
`envset-plugin-redis`

``` js
...
yargs
.command(require('envset-plugin-s3'))
.command(require('envset-plugin-redis'))
...
```

and viola, pluginized and npm published and envset plugin ecosystem created. there should also be a standardization to the envset plugins, for example, they must also be 12-factor, therefore all variables should be read from the environment.

**Comment by goliatone** [comment]
It looks pretty convenient, however, it would like it to have a way for it to be dynamic, i.e. you issue a command and try to load it. There is no penalty for having to preload all plugins if you are only going to use one per invocation and you can install command on demand.  

Currently I use a different tool to handle sharing private `envset` files between a team, it uses S3 to store an encrypted version of the file, and it downloads it on demand. It does the job but is not very user friendly, so I'm thinking about rewriting it to use AWS KMS or keybase. 

I think it could be possible to then create an envset plugin that wraps that library.
